### ISSUE TITLE
Context Responses for Lock Ratio Checkbox

### ISSUE THREAD

**Issue by terracoda** [phetsims/ratio-and-proportion#227]
I recently realized we do not have context responses for the Lock Ratio checkbox.
It is a strange interaction with some complicating scenarios that can knock the hands out of ratio, so this is what I suggest.

Goal is to alert the user that the locked situation is no longer in play, and continue on with business as usual in an unlocked situation. I am not sure if this is the best approach, but it is my first best guess.

Checkbox scenarios | Context Response | Event
-- | -- | --
**Typical scenario**: Checking the Lock Ratio checkbox when hands are at challenge ratio. | Hands locked at challenge ratio. Move hands and observe. | Learner checks the checkbox when it is enabled.
**Edge case 1**: Lock Ratio is checked and then there is a change made to My Challenge. The hands are knocked out of ratio and the checkbox gets unchecked and disabled. | Ratio no longer locked. {{Current challenge {{L to R}}, Hands, {{proximity-to-ratio}} to new challenge ratio.}} | Checkbox is checked and a change to My Challenge knocks hands out of ratio
**Edge case 2**: Lock Ratio checkbox is checked and a learner uses the number shortcuts. The hands likely get knocked out of ratio and the checkbox is unchecked and disabled. | Ratio no longer locked. {{Hands {{even with}} each other, {{proximity-to-ratio}} to challenge ratio.}} | Checkbox is checked and a learner jumps both hands to a position (0-9) that effectively knocks the hands out of ratio unless, of course, challenge ratio is set to 1:1 - 9:9

**Comment by terracoda** [comment]
Note I updated the content in the table after initial post. No changes to the responses - just the supporting text.

**Comment by zepumph** [comment]
Above I added the typical context response based on the checkbox firing. I didn't get to the other two before I ran out of time for the day. I feel ok about the incomplete state for interviews, since this is the main case. @terracoda, please field any weirdness that may come from no context response when automatically unlocking the ratio (both hands number keys or changing the challenge ratio).

**Comment by terracoda** [comment]
I updated the phrase in the table in https://github.com/phetsims/ratio-and-proportion/issues/227#issue-735442323

**Comment by zepumph** [comment]
In the above table, there was no indication of a context response when you uncheck the checkbox, so I added it above. It says "Ratio no longer locked." 

Next I will add the atypical cases that can cause the checkbox to be unchecked.

**Comment by zepumph** [comment]
Edge case 1 is turning out to be very challenging to code up. Notes for myself:

Order of operation for NumberPicker:
1. input fires a listener on the NumberPicker
2. Property track:
  a. That input changes a value
  b.  In the model that value change may or may not unlock the ratio
3. Input listener on Node Track:
  a. AccessibleValueHandler is listener for the end of the interaction to give a context response 
  b. Conditionally, if the model went from locked to unlocked from this stack, adjust the context respones to be edge case 1's description.

Basically I need to know about 2.b from 3.b, which is not currently possible. I want to talk with @jessegreenberg about how I may be able to do this.

**Comment by zepumph** [comment]
I can do edge case 2 though.

**Comment by zepumph** [comment]
Edge case 2 implemented above. The "Ratio no longer locked." is not part of the same utterance as the normal object/context response. I tried to time it so that it aligned like you have above, but it would be good to test on something other than NVDA.

**Comment by terracoda** [comment]
Typical Scenario is good.
![Screen Shot 2020-12-07 at 4 33 27 PM](https://user-images.githubusercontent.com/1902116/101399494-2027b100-38aa-11eb-8840-94b0bbfccef3.png)
![Screen Shot 2020-12-07 at 4 33 39 PM](https://user-images.githubusercontent.com/1902116/101399487-1e5ded80-38aa-11eb-9922-8310e0200b50.png)

**Comment by zepumph** [comment]
Note that Edge case 1 is not yet implemented here.

**Comment by terracoda** [comment]
@zepumph, said:
"Edge case 2 implemented above. The "Ratio no longer locked." is not part of the same utterance as the normal object/context response. I tried to time it so that it aligned like you have above, but it would be good to test on something other than NVDA."

I don't actually hear, "Ratio, no longer locked." when using VO.
The string flashes quickly in the VO box, but does stay long enough to actually be uttered.

@zepumph, is there any reason why you have decided to NOT put the string in the same response, i.e., to prepend it to make a longer context response? From a design perspective, it seems like one context response to me. The learner presses a number key and 3 important things happen all at once, the ratio gets unlocked, the hands are in a new position, and the hands' proximity to ratio has changed. A single context response is what I think we need:

- "Ratio no longer locked. {{Hands {{even with}} each other, {{proximity-to-ratio}} to challenge ratio.}}" 

Hmm, considering the design flip in #245 that happened, I am wondering if a better context response might be:
- {{Ratio no longer locked.}} {{Hands {{qual-distance-region}} each other, both hands {{pa-region}}. 
    - "Ratio no longer locked. Hands, even with each other, both hands at middle."
    - "Ratio no longer locked. Hands, even with each other, both hands in upper region."
    - "Ratio no longer locked. Hands, even with each other, both hands in lower-middle region."
    - "Ratio no longer locked. Hands, even with each other, both hands at bottom."
    - "Ratio no longer locked. Hands, even with each other, both hands on fifth tick mark."
    - "Ratio no longer locked. Hands, even with each other, both hands on tick mark 8."
    - "Ratio no longer locked. Hands, even with each other, both hands on tick mark 4."
   
We can discuss with the design team, if a more specific phrase should be prepended, e.g.:
- Ratio no longer locked at challenge.
- No longer at challenge ratio.
- No longer locked, {{proximity-to-ratio}} challenge ratio.
    - e.g.: No longer locked, {{far from}} challenge ratio.

**Comment by terracoda** [comment]
If we go with a static parameter about the locking part, like this:
- "{{Ratio no longer locked.}} {{Hands {{qual-distance-region}} each other, both hands {{pa-region}}."

Then we assume, the learners has to go back to the hands to find the actual proximity to the current ratio, knowing already that the hands are no longer locked at the challenge ratio.

If we want to actually say everything in the response, we should go with  a dynamic construction in the locking part, as suggested above:
- No longer locked, {{proximity-to-ratio}} challenge ratio. + the other 2 strings

**Comment by terracoda** [comment]
@zepumph, my main question is still about the separate responses.
Is there any specific reason why you have decided to NOT use one compiled response. That seems way simple to me, and less fragile in terms of missing a response.

**Comment by terracoda** [comment]
In https://github.com/phetsims/ratio-and-proportion/issues/227#issuecomment-732496032, @zepumph said:
> 2. Property track:
> a. That input changes a value
> b. In the model that value change may **or may not unlock the ratio**

@zepumph, if ratio is locked, a change in the number spinners will always knock the hands out of ratio, right? There is no "may not unlock" case, here.

**Comment by terracoda** [comment]
Yes, let's discuss tomorrow, glad you marked it with meeting label.

**Comment by zepumph** [comment]
Today @terracoda and I discussed this extensively, mostly we described the complexity of modular mvc coding and why that is challenging for complex description and context responses.

Steps from here:
- [ ] Make the "ratio unlocked" part of the alert assertie
- [ ] Discuss with @jessegreenberg potentially better design patterns for description in these context cases.
- [ ] Potentially @terracoda said these edge cases aren't a deal breaker if we have to publish without. I hope we don't get there.

**Comment by zepumph** [comment]
> Discuss with @jessegreenberg potentially better design patterns for description in these context cases.

@jessegreenberg and I did end up discussing this, and we think that the best way to solve this is to know about all of the description dependencies ahead of time, because for cases like this, they drastically change the way that the code is set up.

If we know that a description needs to know  about the lock ratio checkbox interaction, the number pickers in the accordion box, and the ratio.lockedProperty in the Model from the start, then it will be more obvious that we need a way for the description to exist outside of any individual piece of that. Right now they are all modular to their own components, and are designed now to know about each other. 

To solve this particular case, I will try to hack around the challenge without rewriting each components' description.

**Comment by zepumph** [comment]
After the above commit, this is the order in which I get alerts while highlighting the both-hands interaction while locked (targetRatio not1), and pressing a number key:

![image](https://user-images.githubusercontent.com/6856943/103238120-30094280-48f6-11eb-96ec-be37b3fbc90b.png)


So now the "Ratio no longer locked" is after the assertive alert, so it won't get interrupted. This sounds really good on NVDA. 

- [ ] @terracoda please test edge case 2, as it should be working correctly now.
- [ ] @terracoda please confirm that the normal behavior (first row in initial table is still working as expected).
- [ ] @zepumph still needs to work on edge case 1.

**Comment by zepumph** [comment]
I'm happy to report that the above commits add a "Ratio no longer locked" separate alert right before the context response when changing the my challenge AccordionBox (edge case 1). 

@terracoda! We are ready for a complete review of the issue. Let me know how it goes!.

- [ ] @terracoda to review edge case 1

**Comment by terracoda** [comment]
@zepumph, I wasn't sure the separate responses would work, but they work very nicely in all cases.
I reviewed all three cases.

I found a new issue when reviewing the number key shortcuts. The hands don't move when ratio is 1:1 and the Lock Ratio checkbox is checked and I go to use the number key shortcuts. I'll open a new issue for that,

**Comment by terracoda** [comment]
Awesome work @zepumph, closing.

**Comment by zepumph** [comment]
Thanks! 

>I found a new issue when reviewing the number key shortcuts. The hands don't move when ratio is 1:1 and the Lock Ratio checkbox is checked and I go to use the number key shortcuts. I'll open a new issue for that,

FYI I did a bunch of refactoring related to the code review over in #286, so likely there is a bit of buggy weirdness from that. Thanks for being on the lookout.

**Comment by terracoda** [comment]
Re-opening.

@zepumph, in light of a new response needed as a result of design changes needed for #289, I noticed that for Edge case 2, I think I am getting the wrong response. I see from my previous comments, things may have gotten very confusing. I think I confused things by using strings from what was implemented rather than what I originally requested.

So I am copying an updated table from the design document here.
I need to discuss the updated [Table: Lock Ratio Context Responses](https://docs.google.com/document/d/1HJvvGWYaUv9tlat9S7omQQVng8fGB4FuUhTEM1qx1z8/edit#heading=h.7kac3owqb80b) with you.

I think Edge Case 2 and the new Edge 3 are quite related and can potentially share the same second response.

**Comment by zepumph** [comment]
If this needs a conversation between us, please mark for meeting, otherwise please assign to me for implementation.

**Comment by zepumph** [comment]
Also, just to have over here, in https://github.com/phetsims/ratio-and-proportion/issues/289#issuecomment-754888778 I said (from meeting discussion):

>We decided that when locked and target is 1, pressing "0" in the both hands interaction should not work. In addition to this patch. @terracoda will come up with an alert to say in this case.

- [ ] @terracoda will come up with the alert in this issue.

**Comment by terracoda** [comment]
@zepumph , marking for a quick meeting to discuss the best way to handle edge cases 2 and 3.
I do have a proposed alert designed fro edge case 3 which was brought up in #289.

**Comment by terracoda** [comment]
And a proposed change to that would follow the same pattern as edge case 3.
I think they should both follow the same pattern.

**Comment by terracoda** [comment]
@zepumph, maybe we do not need to meet about his issue. I have a new proposal that aligns edges case 2 and 3, to use already available object-response strings. I outlined 2 other possible approaches to Edge case 2 in the design document in the [Table: Lock Ratio Context Responses](https://docs.google.com/document/d/1HJvvGWYaUv9tlat9S7omQQVng8fGB4FuUhTEM1qx1z8/edit#heading=h.7kac3owqb80b), but I am most interested to know if my 3rd option for Edge Case 2, and my only suggested option for Edge Case 3 are possible. I feel like the learner needs the same object information in these 2 cases.

Copying my preferred solutions from design doc for Edge Case 2 & 3:


Checkbox scenarios | Context Response
-- | --
Edge case 2: Hands are knocked out of ratio via use of number key shortcuts (0-9). | - Ratio no longer locked.  - hands, {{proximity-to-ratio}} to challenge ratio
Edge case 3: My Challenge is N:N, the hands get knocked out of ratio and the checkbox is unchecked and disabled. | New context response needed for issue #289:Can’t jump both hands to zero with ratio locked.hands, {{proximity-to-ratio}} to challenge ratio

**Comment by terracoda** [comment]
@zepumph, please re-mark with meeting label if you think we need to discuss.

**Comment by zepumph** [comment]
Something like 


```js
    onInput: ( knockedOutOfLock, zeroCasePressed ) => {
        this.alertBothHandsContextResponse( knockedOutOfLock, zeroCasePressed );
      }

**Comment by zepumph** [comment]
@terracoda everything was implemented as you wanted to get edge case 2 and 3 in place. I got pretty confused about edge case 2 while doing it though. We will always hear the object response from the both hands interaction before the context response, so by making the second piece the proximity to ratio also (duplicate of the both hands object response), we are really just duplicating that info in a sandwich around the "ratio no longer locked" alert. Is this desirable? 

![image](https://user-images.githubusercontent.com/6856943/104658276-ea2ad900-566e-11eb-8d2e-eb56cc8a1f0a.png)

I feel like it would be a bit nicer to just have the normal, positional context response after "Ratio no longer locked."


Please review on master.

**Comment by terracoda** [comment]
@zepumph, yes, Edge Case 2 and Edge Case 3 are working on MacOS with keyboard.
Thanks for the extra effort on this one.
